cirq-rigetti/cirq_rigetti/quil_input.py

```diff
@@ -289,4 +435,214 @@ def circuit_from_quil(quil: str) -> Circuit:
                 f"Quil instruction {inst} of type {type(inst)} not currently supported in Cirq."
             )
 
+    if len(kraus_model) > 0:
+        noise_model = kraus_noise_model_to_cirq(kraus_model, defined_gates)
+        circuit = circuit.with_noise(noise_model)
+
     return circuit
+
+
+def get_defined_gates(program: Program) -> Tuple[Dict, Dict]:
+    """Get the gate definitions for the program. Will include the default SUPPORTED_GATES, in
+    addition to any gates defined in the Quil
+
+    Args:
+        program: A pyquil program which may contain some DefGates.
+
+    Returns:
+        A dictionary mapping quil gate names to Cirq Gates
+        A dictionary mapping quil gate names to callable parameter transformers
+    """
+    defined_gates = SUPPORTED_GATES.copy()
+    parameter_transformers = PARAMETRIC_TRANSFORMERS.copy()
+    for defgate in program.defined_gates:
+        if defgate.parameters:
+            defined_gates[defgate.name] = defgate_to_cirq(defgate)
+            parameter_transformers[defgate.name] = lambda *args: {
+                p.name: a for p, a in zip(defgate.parameters, args)
+            }
+        else:
+            defined_gates[defgate.name] = MatrixGate(np.asarray(defgate.matrix, dtype=np.complex_))
+    return defined_gates, parameter_transformers
+
+
+def kraus_noise_model_to_cirq(
+    kraus_noise_model: Dict[Tuple[QubitDesignator, ...], List[NDArray[np.complex_]]],
+    defined_gates: Optional[Dict[QubitDesignator, Gate]] = None,
+) -> InsertionNoiseModel:
+    """Construct a Cirq noise model from the provided Kraus operators.
+
+    Args:
+        kraus_noise_model: A dictionary where the keys are tuples of Quil gate names and qubit
+        indices and the values are the Kraus representation of the noise channel.
+        defined_gates: A dictionary mapping Quil gates to Cirq gates.
+    Returns:
+        A Cirq InsertionNoiseModel which applies the Kraus operators to the specified gates.
+    """
+    if defined_gates is None:
+        """
+        Incompatible types in assignment
+            expression has type
+                "         Dict[str, Union[Gate, Callable[..., Gate]]]"
+            variable has type
+                "Optional[Dict[Union[Qubit, QubitPlaceholder, FormalArgument, int], Gate]]"
+        """
+        defined_gates = SUPPORTED_GATES
```

Comment:
I can potentially convert the key to a str just by using `str`, but what should I do for the members of `SUPPORTED_GATES` that are `Callable[..., Gate]`? Can those be invoked without any parameters?

Reply:
I think the type here is simply wrong. The type of `defined_gates` should match supported_gates.

Reply:
Ok. I am going to ignore this type error then.